File: linux/src/core/ClickRateParser.cpp
#include "ClickRateParser.h"

#include <cmath>

namespace QuadClicker {

bool ClickRateParser::tryParsePositive(const QString& s, double& value)
{
    bool ok = false;
    value = s.trimmed().toDouble(&ok);
    return ok && value > 0.0;
}

bool ClickRateParser::buildDelay(double ms,
                                  std::chrono::milliseconds& delay,
                                  QString& error)
{
    if (!std::isfinite(ms)) {
        error = QStringLiteral("Rate is not a finite number.");
        return false;
    }
    if (ms < MinDelayMs) {
        error = QStringLiteral("Rate exceeds maximum — minimum delay is 1 ms (1000 clicks/sec).");
        return false;
    }
    if (ms > MaxDelayMs) {
        error = QStringLiteral("Delay exceeds maximum of 360 minutes.");
        return false;
    }
    delay = std::chrono::milliseconds(static_cast<long long>(ms));
    error.clear();
    return true;
}

bool ClickRateParser::endsWithAny(const QString& t, std::initializer_list<const char*> suffixes)
{
    for (auto s : suffixes) {
        if (t.endsWith(QLatin1String(s))) return true;
    }
    return false;
}

QString ClickRateParser::stripFirstSuffix(const QString& t, std::initializer_list<const char*> suffixes)
{
    for (auto s : suffixes) {
        QLatin1String suf(s);
        if (t.endsWith(suf)) {
            return t.left(t.size() - suf.size()).trimmed();
        }
    }
    return t.trimmed();
}

bool ClickRateParser::tryParse(const QString& text,
                                std::chrono::milliseconds& delay,
                                QString& error)
{
    delay = std::chrono::milliseconds(100);
    error.clear();

    QString t = text.trimmed().toLower();

    if (t.isEmpty()) {
        error = QStringLiteral("Click rate is required.");
        return false;
    }

    // ── Milliseconds: "100ms" ──────────────────────────────────────────────────
    if (t.endsWith(QLatin1String("ms"))) {
        QString num = t.chopped(2).trimmed();
        double ms = 0.0;
        if (tryParsePositive(num, ms)) {
            return buildDelay(ms, delay, error);
        }
        error = QStringLiteral("Millisecond value must be a positive number.");
        return false;
    }

    // ── Minutes: "2m", "2min", "2minutes" ─────────────────────────────────────
    if (endsWithAny(t, {"minutes", "minute", "mins", "min", "m"})
        && !t.contains(QLatin1String("per minute"))
        && !t.endsWith(QLatin1String("/min"))
        && !t.endsWith(QLatin1String("cpm")))
    {
        QString num = stripFirstSuffix(t, {"minutes", "minute", "mins", "min", "m"});
        double mins = 0.0;
        if (tryParsePositive(num, mins)) {
            return buildDelay(mins * 60'000.0, delay, error);
        }
        error = QStringLiteral("Minutes value must be a positive number.");
        return false;
    }

    // ── Seconds: "5s", "5sec", "5seconds" ─────────────────────────────────────
    if (endsWithAny(t, {"seconds", "second", "secs", "sec", "s"})
        && !t.contains(QLatin1String("per second"))
        && !t.endsWith(QLatin1String("/s"))
        && !t.endsWith(QLatin1String("cps")))
    {
        QString num = stripFirstSuffix(t, {"seconds", "second", "secs", "sec", "s"});
        double secs = 0.0;
        if (tryParsePositive(num, secs)) {
            return buildDelay(secs * 1000.0, delay, error);
        }
        error = QStringLiteral("Seconds value must be a positive number.");
        return false;
    }

    // ── Clicks/second: "10/s", "10cps", "10 times per second" ─────────────────
    if (t.endsWith(QLatin1String("/s")) ||
        t.endsWith(QLatin1String("cps")) ||
        t.contains(QLatin1String("times per second")))
    {
        QString num = t;
        num.replace(QLatin1String("times per second"), QString())
           .replace(QLatin1String("/s"),               QString())
           .replace(QLatin1String("cps"),              QString())
           .replace(QLatin1String(" "),                QString());
        double tps = 0.0;
        if (tryParsePositive(num, tps)) {
            return buildDelay(1000.0 / tps, delay, error);
        }
        error = QStringLiteral("Clicks-per-second value must be a positive number.");
        return false;
    }

    // ── Clicks/minute: "600/min", "600cpm", "600 times per minute" ────────────
    if (t.endsWith(QLatin1String("/min")) ||
        t.endsWith(QLatin1String("cpm")) ||
        t.contains(QLatin1String("times per minute")))
    {
        QString num = t;
        num.replace(QLatin1String("times per minute"), QString())
           .replace(QLatin1String("/min"),             QString())
           .replace(QLatin1String("cpm"),              QString())
           .replace(QLatin1String(" "),                QString());
        double tpm = 0.0;
        if (tryParsePositive(num, tpm)) {
            return buildDelay(60'000.0 / tpm, delay, error);
        }
        error = QStringLiteral("Clicks-per-minute value must be a positive number.");
        return false;
    }

    // ── Clicks/hour: "60/h", "60cph", "60 times per hour" ─────────────────────
    if (t.endsWith(QLatin1String("/h")) ||
        t.endsWith(QLatin1String("cph")) ||
        t.contains(QLatin1String("times per hour")))
    {
        QString num = t;
        num.replace(QLatin1String("times per hour"), QString())
           .replace(QLatin1String("/h"),             QString())
           .replace(QLatin1String("cph"),            QString())
           .replace(QLatin1String(" "),              QString());
        double tph = 0.0;
        if (tryParsePositive(num, tph)) {
            return buildDelay(3'600'000.0 / tph, delay, error);
        }
        error = QStringLiteral("Clicks-per-hour value must be a positive number.");
        return false;
    }

    // ── Bare integer/decimal → milliseconds ───────────────────────────────────
    double bare = 0.0;
    if (tryParsePositive(t, bare)) {
        return buildDelay(bare, delay, error);
    }

    error = QStringLiteral("Invalid format. Examples: 100ms  |  5s  |  10/s  |  600/min  |  60/h");
    return false;
}

} // namespace QuadClicker

File: macos/QuadClicker/Core/ClickRateParser.swift
// Core/ClickRateParser.swift
// QuadClicker — macOS

import Foundation

enum ClickRateParser {
    static let minDelayMs: Double = 1.0
    static let maxDelayMs: Double = 360.0 * 60_000.0   // 360 minutes

    /// Parse `text` into a delay TimeInterval (seconds).
    static func parse(_ text: String) -> Result<TimeInterval, String> {
        let t = text.trimmingCharacters(in: .whitespaces).lowercased()

        if t.isEmpty {
            return .failure("Click rate is required.")
        }

        // ── Milliseconds: "100ms" ─────────────────────────────────────────────
        if t.hasSuffix("ms") {
            let num = String(t.dropLast(2)).trimmingCharacters(in: .whitespaces)
            if let ms = parsePositive(num) {
                return buildDelay(ms)
            }
            return .failure("Millisecond value must be a positive number.")
        }

        // ── Minutes: "2m", "2min", "2minutes" ─────────────────────────────────
        if endsWithAny(t, ["minutes", "minute", "mins", "min", "m"])
           && !t.contains("per minute")
           && !t.hasSuffix("/min")
           && !t.hasSuffix("cpm")
        {
            let num = stripFirstSuffix(t, ["minutes", "minute", "mins", "min", "m"])
            if let mins = parsePositive(num) {
                return buildDelay(mins * 60_000.0)
            }
            return .failure("Minutes value must be a positive number.")
        }

        // ── Seconds: "5s", "5sec", "5seconds" ─────────────────────────────────
        if endsWithAny(t, ["seconds", "second", "secs", "sec", "s"])
           && !t.contains("per second")
           && !t.hasSuffix("/s")
           && !t.hasSuffix("cps")
        {
            let num = stripFirstSuffix(t, ["seconds", "second", "secs", "sec", "s"])
            if let secs = parsePositive(num) {
                return buildDelay(secs * 1000.0)
            }
            return .failure("Seconds value must be a positive number.")
        }

        // ── Clicks/second: "10/s", "10cps", "10 times per second" ────────────
        if t.hasSuffix("/s") || t.hasSuffix("cps") || t.contains("times per second") {
            let num = t
                .replacingOccurrences(of: "times per second", with: "")
                .replacingOccurrences(of: "/s", with: "")
                .replacingOccurrences(of: "cps", with: "")
                .trimmingCharacters(in: .whitespaces)
            if let tps = parsePositive(num) {
                return buildDelay(1000.0 / tps)
            }
            return .failure("Clicks-per-second value must be a positive number.")
        }

        // ── Clicks/minute: "600/min", "600cpm", "600 times per minute" ────────
        if t.hasSuffix("/min") || t.hasSuffix("cpm") || t.contains("times per minute") {
            let num = t
                .replacingOccurrences(of: "times per minute", with: "")
                .replacingOccurrences(of: "/min", with: "")
                .replacingOccurrences(of: "cpm", with: "")
                .trimmingCharacters(in: .whitespaces)
            if let tpm = parsePositive(num) {
                return buildDelay(60_000.0 / tpm)
            }
            return .failure("Clicks-per-minute value must be a positive number.")
        }

        // ── Clicks/hour: "60/h", "60cph", "60 times per hour" ────────────────
        if t.hasSuffix("/h") || t.hasSuffix("cph") || t.contains("times per hour") {
            let num = t
                .replacingOccurrences(of: "times per hour", with: "")
                .replacingOccurrences(of: "/h", with: "")
                .replacingOccurrences(of: "cph", with: "")
                .trimmingCharacters(in: .whitespaces)
            if let tph = parsePositive(num) {
                return buildDelay(3_600_000.0 / tph)
            }
            return .failure("Clicks-per-hour value must be a positive number.")
        }

        // ── Bare integer/decimal → milliseconds ───────────────────────────────
        if let bare = parsePositive(t) {
            return buildDelay(bare)
        }

        return .failure("Invalid format. Examples: 100ms  |  5s  |  10/s  |  600/min  |  60/h")
    }

    // ── Helpers ──────────────────────────────────────────────────────────────

    private static func buildDelay(_ ms: Double) -> Result<TimeInterval, String> {
        if !ms.isFinite { return .failure("Rate is not a finite number.") }
        if ms < minDelayMs {
            return .failure("Rate exceeds maximum — minimum delay is 1 ms (1000 clicks/sec).")
        }
        if ms > maxDelayMs {
            return .failure("Delay exceeds maximum of 360 minutes.")
        }
        return .success(ms / 1000.0)
    }

    private static func parsePositive(_ s: String) -> Double? {
        let trimmed = s.trimmingCharacters(in: .whitespaces)
        guard let v = Double(trimmed), v > 0 else { return nil }
        return v
    }

    private static func endsWithAny(_ text: String, _ suffixes: [String]) -> Bool {
        for s in suffixes where text.hasSuffix(s) { return true }
        return false
    }

    private static func stripFirstSuffix(_ text: String, _ suffixes: [String]) -> String {
        for s in suffixes where text.hasSuffix(s) {
            return String(text.dropLast(s.count)).trimmingCharacters(in: .whitespaces)
        }
        return text.trimmingCharacters(in: .whitespaces)
    }
}
